yktrace/src/tir.rs

```diff
@@ -36,18 +36,96 @@ lazy_static! {
+    /// built thus-far in an error case, but since we have no sensible way to stop the tracer at
+    /// the right time, inevitably the trace shoots off into code which has no SIR. In turn this
+    /// invalidates the trace. We only return the trace built so far for now so that we have
+    /// *something* to test in the interim.
+    pub(crate) fn new(trace: &'_ dyn SirTrace) -> Result<Self, Self> {
+        let mut ops = Vec::new();
+        let num_locs = trace.len();
+
+        for blk_idx in 0..num_locs {
+            let loc = trace.loc(blk_idx);
+            let body = match SIR_MAP.get(&DefId::from_sir_loc(loc)) {
+                Some(b) => b,
+                _ => return Err(Self { ops })
+            };
+
+            let shadow_bb_idx_usize = usize::try_from(loc.bb_idx()).unwrap();
+            // Here we use an invariant of the MIR transform to find the user block for a shadow
+            // block. In the blocks vector, first come N shadow blocks, then come N corresponding
+            // user blocks. The debug assertion checks the invariant holds by looking at where the
+            // shadow block returns to after the call to the trace recorder.
+            let user_bb_idx_usize = body.blocks.len() / 2 + shadow_bb_idx_usize;
+            #[cfg(debug_assertions)]
+            match body.blocks[shadow_bb_idx_usize].term {
+                Terminator::Call{ret_bb, ..} => debug_assert!(ret_bb == Some(BasicBlockIndex::try_from(user_bb_idx_usize).unwrap())),
+                _ => panic!("shadow invariant doesn't hold"),
+            }
+
+            // When adding statements to the trace, we clone them (rather than referencing the
+            // statements in the SIR_MAP) so that we have the freedom to mutate them later.
+            ops.extend(
+                body.blocks[user_bb_idx_usize]
+                    .stmts
+                    .iter()
+                    .cloned()
+                    .map(TirOp::Statement)
+            );
+
+            // FIXME: Convert the block terminator to a guard if necessary.
+        }
+        Ok(Self { ops })
+    }
+
+    /// Return the length of the trace measure in operations.
+    pub fn len(&self) -> usize {
+        self.ops.len()
     }
 }
 
 /// A TIR operation. A collection of these makes a TIR trace.
-pub enum _TirOp {
-    // FIXME: implement innards.
+#[derive(Debug)]
+pub enum TirOp {
+    Statement(Statement),
+    // FIXME guards
+}
+
+#[cfg(test)]
+mod tests {
+    // Some work to trace.
+    #[inline(never)]
+    fn work(x: usize, y: usize) -> usize {
+        let mut res = 0;
+        while res < y {
+            res += x;
+        }
+        res
+    }
+
+    use crate::{start_tracing, TracingKind, TirTrace};
+    use test::black_box;
+
+    #[test]
+    fn nonempty_tir_trace() {
+        let tracer = start_tracing(Some(TracingKind::SoftwareTracing));
+        let res = black_box(work(black_box(3), black_box(13)));
+        println!("THis is a teds");
```

Comment:
Huh?

Reply:
LOL! That's debugging code!

Reply:
I think we only need to black box the args too.